### ISSUE TITLE
infra cloud init fails, hanging stack-create

### ISSUE THREAD

**Issue by juddmaltin-dell** [redhat-openstack/openshift-on-openstack#113]
Does all of cloud init fail if a few of the scripts fail?    I'm running the latest rhel*.qcow2

May 13 17:38:00 localhost cloud-init: Loaded plugins: search-disabled-repos
May 13 17:38:01 localhost cloud-init: No package centos-release-openstack-liberty available.
May 13 17:38:01 localhost cloud-init: Error: Nothing to do
May 13 17:38:01 localhost cloud-init: 2016-05-13 17:38:01,254 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/part-010 [1]

Is the above failure fatal?  Is it because I'm indicating a pool-id that might not have a repo with the centos-release-openstack-liberty package in it?

May 13 17:38:01 localhost cloud-init: /var/lib/cloud/instance/scripts/part-011: line 102: os-collect-config: command not found
May 13 17:38:01 localhost cloud-init: 2016-05-13 17:38:01,285 - util.py[WARNING]: Failed running /var/lib/cloud/instance/scripts/part-011 [127]

is the above fatal as a result of the error in part-010?

The install is OK: 

May 13 17:38:45 localhost cloud-init: + notify_success 'OpenShift node has been prepared for running ansible.'
May 13 17:38:45 localhost cloud-init: + curl -i -X POST -H 'X-Auth-Token: d70a0286a8344fa8a6c49d8712da512d' -H 'Content-Type: application/json' -H 'Accept: application/json' http://192.168.190.125:8004/v1/ad3598906ade4a15986530d2d518a77c/stacks/openshift-infra_host-qggq6xaqtdvy/2027bf0d-4f19-4f32-91c8-09aed34062de/resources/wait_handle/signal --data-binary '{"status": "SUCCESS", "reason": "OpenShift node has been prepared for running ansible.", "data": "OpenShift node has been prepared for running ansible."}'
May 13 17:38:45 localhost cloud-init: % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
May 13 17:38:45 localhost cloud-init: Dload  Upload   Total   Spent    Left  Speed
May 13 17:38:45 localhost cloud-init: 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0#015100   157  100     4  100   153     13    505 --:--:-- --:--:-- --:--:--   506#015100   157  100     4  100   153     13    504 --:--:-- --:--:-- --:--:--   506
May 13 17:38:45 localhost cloud-init: HTTP/1.1 200 OK
May 13 17:38:45 localhost cloud-init: Content-Type: application/json; charset=UTF-8
May 13 17:38:45 localhost cloud-init: Content-Length: 4
May 13 17:38:45 localhost cloud-init: X-Openstack-Request-Id: req-916d33fa-1515-496b-b9d1-b1bfd6767d12
May 13 17:38:45 localhost cloud-init: Date: Fri, 13 May 2016 21:38:45 GMT
May 13 17:38:45 localhost cloud-init: null+ exit 0

but cloud-init still seems very sad:

May 13 17:38:45 localhost cloud-init: 2016-05-13 17:38:45,728 - cc_scripts_user.py[WARNING]: Failed to run module scripts-user (scripts in /var/lib/cloud/instance/scripts)
May 13 17:38:45 localhost cloud-init: 2016-05-13 17:38:45,733 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python2.7/site-packages/cloudinit/config/cc_scripts_user.pyc'>) failed
May 13 17:38:45 localhost cloud-init: 2016-05-13 17:38:45,831 - templater.py[WARNING]: Cheetah not available as the default renderer for unknown template, reverting to the basic renderer.
May 13 17:38:45 localhost cloud-init: Cloud-init v. 0.7.6 finished at Fri, 13 May 2016 21:38:45 +0000. Datasource DataSourceOpenStack [net,ver=2].  Up 331.20 seconds

I'm going to run it again without a pool-id, to see if the code just magically does the right thing.

**Comment by jprovaznik** [comment]
Hi Judd,
cloud-init runs all scripts even if some of them fail which is not optimal because then it might happen that the last script signals false-positive successful return code (https://github.com/redhat-openstack/openshift-on-openstack/blob/master/fragments/infra-boot.sh).

What I'm not sure about is why check didn't catch the missing os-collect-config:
https://github.com/redhat-openstack/openshift-on-openstack/blob/master/fragments/infra-boot.sh#L229

This check was added exactly for making sure that os-collect-config is present to avoid the situation you hit.

**Comment by jprovaznik** [comment]
If you hit it again, could you please attach /var/log/cloud-init(-output).log?

**Comment by juddmaltin-dell** [comment]
Failed again.

Here are the params and cloud-init.log attached.  THANK YOU!  I'm totally blocked by this, and I'm actively trying to work it out so I can progress.

| links                 | http://192.168.190.125:8004/v1/c1b740d8571f4cb5aa66f8ddcfdec015/stacks/openshift/3e7ce435-7c5a-4905-a5d4-86bbf6a448f5 (self) |                           [22/1163]
| notification_topics   | []                                                                                                                           |
| parameters            | {                                                                                                                            |
|                       |   "OS::project_id": "c1b740d8571f4cb5aa66f8ddcfdec015",                                                                      |
|                       |   "node_docker_volume_size_gb": "25",                                                                                        |
|                       |   "openshift_nodes_min": "1",                                                                                                |
|                       |   "rhn_pool": "*****_",                                                                                                      |
|                       |   "os_password": "**_**_",                                                                                                   |
|                       |   "flavor": "m1.medium",                                                                                                     |
|                       |   "master_server_group_policies": "[u'affinity']",                                                                           |
|                       |   "deploy_registry": "True",                                                                                                 |
|                       |   "rhn_password": "**_***",                                                                                                  |
|                       |   "os_username": "",                                                                                                         |
|                       |   "domain_name": "example.com",                                                                                              |
|                       |   "loadbalancer_type": "neutron",                                                                                            |
|                       |   "openshift_nodes_max": "5",                                                                                                |
|                       |   "ssh_user": "cloud-user",                                                                                                  |
|                       |   "master_docker_volume_size_gb": "25",                                                                                      |
|                       |   "os_tenant_name": "",                                                                                                      |
|                       |   "node_count": "4",                                                                                                         |
|                       |   "master_hostname": "openshift-master",                                                                                     |
|                       |   "deployment_type": "openshift-enterprise",                                                                                 |
|                       |   "OS::stack_name": "openshift",                                                                                             |
|                       |   "autoscaling": "False",                                                                                                    |
|                       |   "master_count": "3",                                                                                                       |
|                       |   "os_region_name": "",                                                                                                      |
|                       |   "deploy_router": "True",                                                                                                   |
|                       |   "rhn_username": "dellcloudsol",                                                                                            |
|                       |   "node_hostname": "openshift-node",                                                                                         |
|                       |   "internal_subnet": "192.168.0.0/24",                                                                                       |
|                       |   "skip_ansible": "False",                                                                                                   |
|                       |   "openshift_ansible_git_url": "",                                                                                           |
|                       |   "infra_hostname": "infra",                                                                                                 |
|                       |   "skip_dns": "False",                                                                                                       |
|                       |   "OS::stack_id": "3e7ce435-7c5a-4905-a5d4-86bbf6a448f5",                                                                    |
|                       |   "extra_repository_urls": "[]",                                                                                             |
|                       |   "os_auth_url": "",                                                                                                         |
|                       |   "external_network": "nova",                                                                                                |
|                       |   "loadbalancer_image": "",                                                                                                  |
|                       |   "ssh_key_name": "key_name",                                                                                                |
|                       |   "lb_hostname": "lb",                                                                                                       |
|                       |   "openshift_ansible_git_rev": "",                                                                                           |
|                       |   "openshift_sdn": "openshift-sdn",                                                                                          |
|                       |   "dns_nameserver": "[u'8.8.4.4', u'8.8.8.8']",                                                                              |
|                       |   "server_image": "rhel72"                                                                                                   |
|                       | }                                                                                                                            |
| parent                | None                                                                                                                         |
| stack_name            | openshift                                                                                                                    |
| stack_owner           | None                                                                                                                         |
| stack_status          | CREATE_IN_PROGRESS                                                                                                           |
| stack_status_reason   |                                                                                                                              |
| stack_user_project_id | 9c985405ca194bc398eda5b9ff8c2241                                                                                             |
| tags                  | null                                                                                                                         |
| template_description  | Deploy Atomic/OpenShift 3 on OpenStack.                                                                                      |
| timeout_mins          | 120                                                                                                                          |
| updated_time          | None                                                                                                                         |
+-----------------------+------------------------------------------------------------------------------------------------------------------------------+
2016-05-16 16:09:05 [openshift]: CREATE_IN_PROGRESS  Stack CREATE started
2016-05-16 16:09:05 [ansible_keys]: CREATE_IN_PROGRESS  state changed

....

Gist of the cloud-init.log:  https://gist.github.com/juddmaltin-dell/cc84f450e3778062abd8d257affcdcf3

**Comment by juddmaltin-dell** [comment]
The test here fails with a false positive:
https://github.com/redhat-openstack/openshift-on-openstack/blob/master/fragments/infra-boot.sh#L162

because it's called with --no-block

It fails because the os-collect-config service was installed **without** the os-collect-config package being installed.

os-collect-config lives in the **rhel-7-server-openstack-8-director-rpms** repo on another box of mine.  But it does not live in any repos available to my OpenShift NFR pool-id.

The repos that exist from the OpenShift NFR pool-id are: 

[root@openshift-infra ~]# yum repolist disabled|grep openstack
rhel-7-server-openstack-7.0-tools-debug-rpms/7Server/x86_64      Red Hat OpenSta
rhel-7-server-openstack-7.0-tools-rpms/7Server/x86_64            Red Hat OpenSta
rhel-7-server-openstack-7.0-tools-source-rpms/7Server/x86_64     Red Hat OpenSta
rhel-7-server-openstack-8-tools-debug-rpms/7Server/x86_64        Red Hat OpenSta
rhel-7-server-openstack-8-tools-rpms/7Server/x86_64              Red Hat OpenSta
rhel-7-server-openstack-8-tools-source-rpms/7Server/x86_64       Red Hat OpenSta

Can we attach multiple pool-ids so the os-collect-config is available?

**Comment by juddmaltin-dell** [comment]
Re-Opened because it's still maybe a bug.

**Comment by jprovaznik** [comment]
Thanks for deep investigation. Yes, this is a bug and the "systemctl is-enabled os-collect-config" is insufficient - the reason is that the service file is not part of the package but is always created here:
 https://github.com/redhat-openstack/openshift-on-openstack/blob/master/collect-config-setup/fragments/start_config_agent.sh#L13

Yum returns success event if some packages fail to install (https://github.com/redhat-openstack/openshift-on-openstack/blob/master/collect-config-setup/fragments/install_config_agent_yum.sh#L19) - this is not the first time I hit this "feature" which I really don't like :/. As a first step we should install os-collect-config in a separate "yum install".

**Comment by juddmaltin-dell** [comment]
Is it possible to include more than one pool?  The OpenShift NFR pool is not reliably responding with _-director-_ repos which include the os-collector-config package.

**Comment by juddmaltin-dell** [comment]
Thank you, fixed.
